vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+            <<"tree size: "<<b.nodes.size()<<endl
+            <<"max number of unique labels: "<<b.max_num_labels<<endl
+            <<"learn at leaf: "<<b.learn_at_leaf<<endl
+            <<"num of dream operations per example: "<<b.dream_repeats<<endl
+            <<"current_pass: "<<b.current_pass<<endl
+            <<"oas: "<<b.oas<<endl;
+    }
+
+
+    //rout based on the prediction
+    inline uint32_t descent(node& n, const float prediction)
+    { 
+        //prediction <0 go left, otherwise go right
+        if(prediction < 0){
+            n.nl++; //increment the number of examples routed to the left.
+            return n.left;
+        }
+        else{ //otherwise go right.
+            n.nr++; //increment the number of examples routed to the right.
+            return n.right; 
+        }
+    }
+
+    //return the id of the example and the leaf id (stored in cn)
+    inline int random_sample_example_pop(memory_tree& b, uint32_t& cn)
+    {
+        cn = 0; //always start from the root:
+        while (b.nodes[cn].internal == 1)
+        {
+            float pred = 0.;   //deal with some edge cases:
+            if (b.nodes[cn].nl < 1) //no examples routed to left ever:
+                pred = 1.f; //go right.
+            else if (b.nodes[cn].nr < 1) //no examples routed to right ever:
+                pred = -1.f; //go left.
+            else if ((b.nodes[cn].nl >= 1) && (b.nodes[cn].nr >= 1))
+                pred = merand48(b.all->random_state) < (b.nodes[cn].nl*1./(b.nodes[cn].nr+b.nodes[cn].nl)) ? -1.f : 1.f;
+            else{
+                cout<<cn<<" "<<b.nodes[cn].nl<<" "<<b.nodes[cn].nr<<endl;
+                cout<<"Error:  nl = 0, and nr = 0, exit...";
+                exit(0);
+            }
+            
+            if (pred < 0){
+                b.nodes[cn].nl--;
+                cn = b.nodes[cn].left; 
+            }
+            else{
+                b.nodes[cn].nr--;
+                cn = b.nodes[cn].right;
+            }
+        }
+
+        if (b.nodes[cn].examples_index.size() >= 1){
+            int loc_at_leaf = int(merand48(b.all->random_state)*b.nodes[cn].examples_index.size());
+	        uint32_t ec_id = b.nodes[cn].examples_index[loc_at_leaf];
+            pop_at_index(b.nodes[cn].examples_index, loc_at_leaf); 
+            return ec_id;
+        }
+        else    
+            return -1;
+    }
+
+
+    //train the node with id cn, using the statistics stored in the node to
+    //formulate a binary classificaiton example.
+    float train_node(memory_tree& b, single_learner& base, example& ec, const uint32_t cn)
+    {
+        //predict, learn and predict
+        //note: here we first train the router and then predict.
+        MULTICLASS::label_t mc;
+        uint32_t save_multi_pred;
+        MULTILABEL::labels multilabels;
+        MULTILABEL::labels preds;
+        if (b.oas == false){
+            mc = ec.l.multi;
+            save_multi_pred = ec.pred.multiclass;
```

Comment:
It's a bit odd to save the prediction when doing training.  Maybe save it once before all training operations?

Reply:
I was just trying to temporally save the label information, and after base.predict is called, I resigned it back (this seems the pattern that I observed in other files like recall tree and Lom tree?)